src/full/Agda/TypeChecking/Rules/Application.hs

```diff
@@ -1044,23 +1044,23 @@ inferOrCheckProjApp e o ds args mt = do
         [ "  principal arg " <+> prettyTCM arg
         , "  has type "      <+> prettyTCM ta
         ]
-      inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta
+      inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta Nothing
 
 -- | Same arguments 'inferOrCheckProjApp' above but also gets the position,
 --   value and type of the principal argument.
 inferOrCheckProjAppToKnownPrincipalArg ::
   A.Expr -> ProjOrigin -> List1 QName -> A.Args -> Maybe (Comparison, Type) ->
-  Int -> Term -> Type -> TCM (Term, Type, CheckedTarget)
-inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta = do
+  Int -> Term -> Type -> Maybe (Args, Type) -> TCM (Term, Type, CheckedTarget)
+inferOrCheckProjAppToKnownPrincipalArg e o ds args mt k v0 ta mat = do
   let cmp = caseMaybe mt CmpEq fst
-      postpone b = do
+      postpone b at = do
         tc <- caseMaybe mt newTypeMeta_ (return . snd)
-        v <- postponeTypeCheckingProblem (CheckProjAppToKnownPrincipalArg cmp e o ds args tc k v0 ta) b
+        v <- postponeTypeCheckingProblem (CheckProjAppToKnownPrincipalArg cmp e o ds args tc k v0 ta at) b
         return (v, tc, NotCheckedTarget)
   -- ta should be a record type (after introducing the hidden args in v0)
-  (vargs, ta) <- implicitArgs (-1) (not . visible) ta
+  (vargs, ta) <- caseMaybe mat (implicitArgs (-1) (not . visible) ta) return
```

Comment:
Doesn't hurt to insert a comment here to explain what is going on...

Reply:
Agreed. I added comments and also referenced issue #4924 for context.

Reply:
Thanks!